synapse/handlers/register.py

```diff
@@ -425,6 +434,9 @@ async def _join_rooms(self, user_id: str):
 
                 # Send the invite, if necessary.
                 if requires_invite:
+                    # If an invite is required, there must be a auto-join user ID.
+                    assert self.hs.config.registration.auto_join_user_id
```

Comment:
This is [asserted in the config code](https://github.com/matrix-org/synapse/blob/f1de4bb58b5b68b0fbb0033bd42fa30df0944281/synapse/config/registration.py#L166-L173) but mypy doesn't know that.